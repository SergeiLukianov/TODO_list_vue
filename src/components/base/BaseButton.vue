<template>
  <div>
    <button
      :class="customStyleClass"
      @click="onClick">{{caption}}</button>
  </div>
</template>

<script>
export default {

  props: {
    caption: {
      type: String,
      default: 'Submit'
    },

    customStyleClass: {
      type:String,
      default: '',
    },
  },

  methods: {
    onClick () {
      this.$emit('click')
    }
  }
}
</script>
