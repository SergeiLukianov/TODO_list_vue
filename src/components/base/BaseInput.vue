<template>
  <div>
    <input
      :placeholder="placeholder"
      :value="value"
      @keydown.enter="onEnter"
      @keydown.esc="onEsc"
      @input="onInput"/>
  </div>
</template>

<script>
export default {
  name: 'BaseInput',

  props: {
    placeholder: {
      type: String,
      default: ''
    },

    value: {
      type: String,
      default: '',
    },
  },

  methods: {
    onInput (event) {
      this.$emit('input', event.target.value)
    },

    onEnter (event) {
      this.$emit('enter', event.target.value)
    },

    onEsc() {
      this.$emit('esc')
    },

  }
}
</script>

<style scoped>

</style>
