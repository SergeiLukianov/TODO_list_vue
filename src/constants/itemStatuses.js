export const OPEN_STATUS = 'open'
export const DONE_STATUS = 'done'
