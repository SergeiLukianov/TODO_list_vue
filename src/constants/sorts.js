const options = {
  recordsAsc: 'Text ASC',
  recordsDesc: 'Text DESC',
  creationDateAsc: 'Date creation ASC',
  creationDateDesc: 'Date creation DESC',
  dueDateAsc: 'Due date ASC',
  dueDateDesc: 'Due date DESC',
};

export default {
  sortOptions: options,
  sortOptionsValues: options.keys(),
}
