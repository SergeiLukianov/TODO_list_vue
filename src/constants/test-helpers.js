const mockState = (store = {}) => JSON.parse(JSON.stringify(store));

const mockGetters = (getters = {}) => Object.keys(getters)
  .reduce((reducedGetters, name) => ({
    ...reducedGetters,
    [name]: jest.fn(),
  }), {});

const mockAction = () => jest.fn(() => Promise.resolve());
const mockActions = (actions = {}) => Object.keys(actions).reduce((mockedActions, name) => ({
  ...mockedActions,
  [name]: mockAction(),
}), {});

const mockMutations = (mutations = {}) => Object.keys(mutations).reduce((mockedMutations, name) => ({
  ...mockedMutations,
  [name]: jest.fn(),
}), {});

export function mockModules(modules) {
  return Object.keys(modules).reduce((mockedModules, name) => {
    let nextModule = modules[name];

    return {
      ...mockedModules,
      [name]: {
        namespaced: nextModule.namespaced,
        state: mockState(nextModule.state),
        getters: mockGetters(nextModule.getters),
        actions: mockActions(nextModule.actions),
        mutations: mockMutations(nextModule.mutations),
      }
    };
  }, {});
}

export function resetModules(modules) {
  return Object.keys(modules).reduce((resetedModules, nextModuleName) => ({
    ...resetedModules,
    [nextModuleName]: {
      ...modules[nextModuleName],
      state: JSON.parse(JSON.stringify(modules[nextModuleName].state || {})),
    },
  }), {});
}

function getNamespacedKeys(obj, moduleName, isNamespaced) {
  return Object.keys(obj).reduce((results, key) => ({
    ...results,
    [key]: (isNamespaced ? `${moduleName}/${key}` : key),
  }), {});
}

export function getActionTypes(moduleConfig) {
  const moduleName = Object.keys(moduleConfig)[0];
  const module = moduleConfig[moduleName];
  const isNamespaced = module.namespaced;

  return getNamespacedKeys(module.actions, moduleName, isNamespaced);
}

export function getGetterTypes(moduleConfig) {
  const moduleName = Object.keys(moduleConfig)[0];
  const module = moduleConfig[moduleName];
  const isNamespaced = module.namespaced;

  return getNamespacedKeys(module.getters, moduleName, isNamespaced);
}
