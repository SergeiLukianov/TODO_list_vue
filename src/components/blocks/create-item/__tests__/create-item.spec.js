import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'

import CreateItem from '../create-item'
import BaseInput from '../../../base/BaseInput'
import BaseButton from '../../../base/BaseButton'
import { mockModules } from '../../../../constants/test-helpers'
import ItemsModule from '../../../../store/modules/items'




describe('CreateItem', () => {
  let wrapper;
  let mockedModules;
  let store;

  const renderWrapper = () => {
    mockedModules = mockModules({ ItemsModule })

    store = new Vuex.Store({
      modules: mockedModules
    })

    let localVue = createLocalVue();
    localVue.component('base-input', BaseInput)
    localVue.component('base-button', BaseButton)

    wrapper = mount(CreateItem, {
      localVue,
      children: [BaseInput, BaseButton],
      store,
    })
  }

  beforeEach(() => {
    renderWrapper()
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('renders', () => {
    expect(wrapper.html()).toMatchSnapshot()
  });

  it('should not create item if text is empty', () => {
    wrapper.vm.createItem()

    //createNewItem action not invoked
    expect(mockedModules.ItemsModule.actions.createNewItem.mock.calls).toEqual([]);
  });

  it('should create item and reset value if text is not empty', () => {
    const value = 'some value'
    wrapper.vm.value = value
    wrapper.vm.createItem()

    expect(mockedModules.ItemsModule.actions.createNewItem.mock.calls[0][1]).toBe(value);
    expect(wrapper.vm.value).toBe('')
  });
})
