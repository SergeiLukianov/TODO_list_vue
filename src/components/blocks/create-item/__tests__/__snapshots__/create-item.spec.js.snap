// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateItem renders 1`] = `
<div class="createItem">
    <div class="customInput"><input placeholder="New task..." class="customInput"></div>
    <div><button class="addItemButton">ADD</button></div>
</div>
`;
